City-of-Helsinki/helsinki-paatokset
# UHF-8284: Platform V3 updates

# [UHF-8284](https://helsinkisolutionoffice.atlassian.net/browse/UHF-8284)

## What was done
Updates all shared modules, project base and project tools to use Platform V3.

## How to install
* Start from `develop` (make sure it's up-to-date) and install content to test with:
  * `git checkout develop`
  * `make new`
* Generate basic test content and pull in integration data:
  * `drush en paatokset_test_content -y;drush pmu paatokset_test_content -y`
  * `drush migrate-import ahjo_decisionmakers:all;drush migrate-import ahjo_decisionmakers:all_sv`
  * `drush migrate-import ahjo_trustees:all;drush migrate-import ahjo_trustees:all_sv`
  * `drush migrate-import ahjo_meetings:latest;drush migrate-import ahjo_decisions:latest`
  * `drush migrate-import paatokset_meeting_videos --update`
* Run TPR migrations so we can check how the upgrade path workrs:
  * `drush migrate-import tpr_unit --limit=100;drush migrate-import tpr_service --limit=100`
* Fetch specific cases and decisions to test with:
```
drush ap:update organization U3212002040VH1 -v
drush ap:update cases hel-2014-000497 -v
drush ap:update cases hel-2016-012511 -v
drush ap:update decisions b4b8e27f-9763-4fd0-810c-96ada6c1c4b7 -v
drush ap:update decisions 2b1c3a6c-f0f2-411e-b6f0-23b56cd0e0ea -v
drush ap:update decisions afe79ec8-328f-414d-85c6-35de280ef0a9 -v
drush ap:update decisions 0e036367-174a-47fb-8ac5-37f923981c16 -v
drush ap:update decisions ffe61d01-9ecb-47b7-a5b3-4eaf1ebe598d -v
drush ap:update decisions fb53d9ca-efaa-4aa2-b417-fe09efeea58a -v
drush ap:update decisions 5940cd98-f540-40b1-947f-c76ff233b93e -v
drush ap:update decisions 20af8888-1394-4c84-96ac-43f78ff8ff4c -v
```
* Get motions and update decision data, then fetch missing cases and decisions:
  - `drush ap:gm -v;drush ap:ud -v`
  - `drush queue:run ahjo_api_retry_queue -v`
  - `drush ap:ud --logic=case -v;drush ap:ud --logic=meeting -v;`
* Add some council meeting minutes of discussion PDFs and link them to meetings:
  - https://helsinki-paatokset.docker.so/fi/media/add/minutes_of_the_discussion
  - Check that they appear on the site: https://helsinki-paatokset.docker.so/fi/paattajat/kaupunginvaltuusto/keskustelupoytakirjat
* Add some custom PDF lists. Reference some content from the production site: https://paatokset.hel.fi/fi/paattajat/rakennuslupien-viranhaltijapaattajat and https://paatokset.hel.fi/fi/paatoksenteko/kaupunginmuseon-lausunnot
  *  Create PDF categories: https://helsinki-paatokset.docker.so/fi/admin/structure/taxonomy/manage/pdf_minutes_category/overview
  * Create custom minutes of discussion media entities: https://helsinki-paatokset.docker.so/fi/media/add/pdf_minutes_of_discussion
  * Add pages under to the "Päätöksenteko" menu item with PDF listing paragraphs: https://helsinki-paatokset.docker.so/fi/node/add/page
  * For at least one of these, add an Accordion element so we can check for conflicts between tabbed content and accordions
* Index content: `drush sapi-i;drush sapi-rt`
* Create a database dump so you can back up the current situation, in case something goes wrong during the upgrades: `make drush-create-dump`

## Install Platform V3 upgrades:
- Read the platform v3 update documentation: https://github.com/City-of-Helsinki/drupal-helfi-platform-config/blob/3.x/documentation/update.md#update-instructions
- Checkout this branch: `git checkout UHF-8284-platform-v3-updates`
  - If you get `unable to unlink old ...` errors on theme assets or setting.php files, you might need to run some variation of: `chmod ug+w public/sites/default` or `chown`, `chgrp` commands on them.
- Run `composer install` and then `make shell`
- Run `drush cr`, `drush helfi:platform-config:update-database` and re-import configuration `drush cim -y` 
  - If the import fails because of `field_hero_design` or `field_dm_org_above` , run: `drush config-delete field.storage.paragraph.field_hero_design;drush config-delete field.storage.node.field_dm_org_above;drush cr;drush cim -y;drush config-delete field.storage.paragraph.field_banner_design -y` 
  - If the configuration import fails for some other reason, run `drush cr` and `drush cim -y`
- If you run `drush cex` (answer no to actually modifying the files), you'll see some configs are still left unimported. Go to https://helsinki-paatokset.docker.so/fi/admin/config/development/configuration/ignore and only leave these two lines:
```
system.site:page.front
easy_breadcrumb.settings:home_segment_title
```` 
- Run `drush cim -y` again. Now https://helsinki-paatokset.docker.so/fi/admin/config/development/configuration/ignore should show the previous state it was in and `drush cex` should no longer show unexported configurations.

## How to test
* [x] Upgrade path goes well and no configuration remains unimported
* [x] Run `drush updb -y`. The updates should go through properly. Run `drush cim -y` and then `drush cex -y`. There shouldn't be any unexported configs.
* [x] Check that nothing weird is going on with the changed configuration files
* [x] Code changes follows our standards
* [ ] Check that the decisions search works: https://helsinki-paatokset.docker.so/fi/asia 
* [x] Decision page layout works:
  * Content areas and voting results: https://helsinki-paatokset.docker.so/fi/asia/hel-2016-012511/b4b8e27f-9763-4fd0-810c-96ada6c1c4b7
  * Decision history dropdown: https://helsinki-paatokset.docker.so/fi/asia/hel-2014-000497
  * Also check out some random decisions from https://helsinki-paatokset.docker.so/fi/admin/content/decisions
* [x] Check that the frontpage hero and meeting calendar works: https://helsinki-paatokset.docker.so/fi/
* [x] Check that the mmeeting calendar works: https://helsinki-paatokset.docker.so/fi/kokouskalenteri
* [x] The decisionmakers listing and search should work as intended: https://helsinki-paatokset.docker.so/fi/kokouskalenteri
* [ ] Decisionmaker page content works:
  * https://helsinki-paatokset.docker.so/fi/paattajat/kaupunginvaltuusto
  * https://helsinki-paatokset.docker.so/fi/paattajat/U3212002040VH1
* [x] Tabbed navigation pages work:
  * https://helsinki-paatokset.docker.so/fi/paattajat/kaupunginvaltuusto/asiakirjat
  * https://helsinki-paatokset.docker.so/fi/paattajat/kaupunginvaltuusto/keskustelupoytakirjat
  * https://helsinki-paatokset.docker.so/fi/kuulutukset-ja-ilmoitukset
  * The custom PDF pages you created. The tabbed navigation shouldnt cause conflicts on pages with accordion elements
* [x] Test that the speaking turns integration works: `drush migrate:import datapumppu_statements -v --no-progress`
* [x] Edit a council members profile page and check that all fields, initiatives, speaking turns etc are displayed correctly: https://helsinki-paatokset.docker.so/fi/paattajat/arhinpa

[UHF-8284]: https://helsinkisolutionoffice.atlassian.net/browse/UHF-8284?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
